### ISSUE TITLE
Palemoon PulseAudio integration implementation

### ISSUE THREAD

**Issue by JStyle21** [MoonchildProductions/Pale-Moon#222]
Hi,

There are several changes and improvements palemoon can take from firefox and\or from implementing proper integration of pulse audio.

For more info take a look at this thread:
https://forum.palemoon.org/viewtopic.php?f=37&t=9934

**Comment by Axiomatic-** [comment]
I will take this. There are a few bugzilla numbers to post here; however, some I find erroneous so I plan to exclude those. (Will edit this post with the bug numbers, after I deem what is valid.)

**Comment by wolfbeast** [comment]
@Axiomatic- Have you have any time to look at this yet?

**Comment by Axiomatic-** [comment]
Yeah, it seems that the back end changes for PulseAudio have been done already, it is just a matter of using them.

See: https://github.com/Axiomatic-/Pale-Moon/tree/PulseAudio-work (enables PulseAudio if detected through libcubeb otherwise fallback to ALSA.)

Untested, and I am not sure how this plays with the OSS option. I will start a build tonight.

**Comment by Axiomatic-** [comment]
Compiles, and works for me @JStyle21 mind testing this?

**Comment by JStyle21** [comment]
@Axiomatic- I will gladly check it out and update, will take a day or two.
Hopefully tomorrow :)

**Comment by Axiomatic-** [comment]
@JStyle21 have you gotten a chance to test this? More specifically does it work with KMix for you?

**Comment by JStyle21** [comment]
@Axiomatic
Sorry with the end of the year celebrations and all i've just got around to it now.
Stupid question but do i need to use the
"ac_add_options --enable-pulseaudio"
to test it or not?

**Comment by Axiomatic-** [comment]
@JStyle21 no worries. You can remove ```--enable-pulseaudio``` from you .mozconfig. If you have PulseAudio installed it should detect to use it at run time.

**Comment by JStyle21** [comment]
@Axiomatic
I've complied the latest version from the master branch (26).
It's still the same as it was without the "--enable-pulseaudio" option it has no sound and Kmix doesn't recognize it.
With the option it's shown as a buffered stream just like in the info (pics) i've provided in my original thread(https://forum.palemoon.org/viewtopic.php?f=37&t=9934).

Also apart from the precompiled caching issue while packaging the build which was conveniently solved by adding "--disable-precompiled-startupcache" i didn't make any other changes to my build since last time.

**Comment by Axiomatic-** [comment]
Yeah, that is what I expected. We can keep this bug open for that; however, the main point was wanting to pick via run time whether to use PulseAudio or ALSA. Thanks for testing. :+1:

**Comment by h1z1** [comment]
Follow up to a chat I had on IRC.  I had to recompile PM without pulseaudio as it's enabled by default in released installs.  I can't stand pulse to the point I use apulse as a wrapper.   Not everyone uses pulse anyway.   PM is the only application I've had a problem with.

Specifically PM would crash with the following whenever any audio was played.  It did not affect flash at all.

    [VS] {full} VdpPresentationQueueDisplay presentation_queue=9, surface=11, clip_width=0, clip_height=0,
           earliest_presentation_time=0

    WARNING: pipe error (3): Connection reset by peer: file /home/trava90/PaleMoon/pmsrc/ipc/chromium/src/chrome/common/ipc_channel_posix.cc, line 442
    
    ###!!! [Child][SyncChannel] Error: Channel error: cannot send/recv
    
    Segmentation fault (core dumped)
    ###!!! [RPCChannel][Child][/home/trava90/PaleMoon/pmsrc/ipc/glue/RPCChannel.cpp:98]    Assertion (!mPendingReply) failed.  should not be waiting for a reply (triggered by rpc)
    RPCChannel 'backtrace':
    remote RPC stack guess: 0
    deferred stack size: 0
    out-of-turn RPC replies stack size: 0
    Pending queue size: 0, front to back:
    ###!!! ABORT: should not be waiting for a reply: file /home/trava90/PaleMoon/pmsrc/ipc/glue/RPCChannel.cpp, line 534
    ###!!! ABORT: should not be waiting for a reply: file /home/trava90/PaleMoon/pmsrc/ipc/glue/RPCChannel.cpp, line 534

Please for the sake of sanity, do not enable pulse by default.  There actually are distributions besides Ubuntu and do not use pulse.

Thanks

**Comment by Axiomatic-** [comment]
I usually run without PulseAudio (and systemd) so I understand that PulseAudio is not installed on every Linux distribution; however, the implementation currently only uses PulseAudio if it is installed. If you do not have PulseAudio installed it _should_ fall back to using ALSA. In the future I could add a preference that the user can set to force the use of ALSA (even if they have PulseAudio installed) if that seems like a thing people would want. 

I will attempt at reproducing your crash (should be later tonight, or sometime tomorrow). If would would kindly excuse my ignorance I do not quite understand if you have PulseAudio installed or not, if you could let me know that would be great. (I am assuming you do not have PulseAudio installed, as I did test with PulseAudio if I recall correctly.)

**Comment by wolfbeast** [comment]
If I understand correctly, @h1z1 doesn't have pulseaudio installed, but uses a wrapper around a different module that emulates pulseaudio. Since the emulation is advertised as "pulseaudio present", Pale Moon will use it. I'm afraid we can't really provide support for that kind of setup and that is likely the cause of the crash (expecting pulseaudio but finding a wrapper, and the wrapper not providing the expected responses or being an incomplete emulation).
If you can just not use the wrapper, it *should* fall back to ALSA instead, automatically.

**Comment by h1z1** [comment]
Curious how pulse is "detected"?  I'm using CentOS 6 and 7, pulse is unfortunately installed on both.   I understand you may not support asound, my concern is the browser crashes entirely.  That can't possibly be what was intended.   If the audio fails to initialize PM should fail gracefully - preferably with an error sound can't be initialized, and continue.

Regarding the wrapper itself, I don't have problems with any other application including Skype, nor Firefox.

**Comment by Axiomatic-** [comment]
I had some time, so I poked at this today. Firefox thinks PulseAudio is installed when ran with apulse, so it tries to call library functions which are not implemented by apulse, then crashes. I am going to go ahead, and assume we should handle that more gracefully like you said, rather than crash. (Again no ETA when I will be able to work up a fix, sorry.)

To add from the apulse README.md: 
> The main objective, working Skype test call, is reached. I don't have any plans for further development.

So the missing library functions is probably expected, the current workaround is either to uninstall PulseAudio and let Pale Moon select ALSA, or run Pale Moon with PulseAudio installed, which will make Pale Moon use PulseAudio.

**Comment by Lesik** [comment]
Recently, [Mozilla decided to make PulseAudio a mandatory dependency of Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1247056). As PulseAudio is not without controversy, you can imagine that this drives a lot of users away from Firefox. I therefore ask everybody in the Pale Moon community to keep PulseAudio integration an _optional_ addition to Pale Moon instead of a mandatory dependency.

**Comment by wolfbeast** [comment]
I don't see a reason to make it mandatory.
Especially if the pulseaudio developer themselves have stated that no further development or maintenance is planned, it'd be rather daft to lock oneself into it as a dependency (unless an implementation would be near-perfect, which is clearly not the case here).

@Axiomatic- Since we've moved forward a lot here since March, can you check the current state of things on master and let me know where we stand now as regards our support and/or crash vs no crash vs handling ot more gracefully now or not? If you no longer have time, let me know and I'll remove assignment for someone else to pick this up.

**Comment by wolfbeast** [comment]
Removed assignment. If anyone wants to address the previous comment here, please do.

**Comment by h1z1** [comment]
Regardless of what you stated in #817, the issue is not with compiling.  PM crashes when Pulseaudio fails to initialize.   That is with release builds.

**Comment by wolfbeast** [comment]
As far as I understood this crash only occurs in the case an incomplete wrapper is used that does not offer the full pulseaudio API (as a shim). Either the shim has to be made complete, or you should not use the wrapper but actual pulseaudio instead where this isn't a problem.

You can't expect the browser to test every function in an API if support is found before using it. That's not how API programming works... If we find pulseaudio support we must be able to assume the implementation of it is complete.

**Comment by h1z1** [comment]
Funny thing is PM's the only application to crash entirely when Pulse fails to initialize.    

I've already forked it.

**Comment by wolfbeast** [comment]
If you have a sane improvement that makes our code more robust in this corner case, I'd certainly be interested in learning what you did :)

**Comment by wolfbeast** [comment]
@trav90 what is our status for this in UXP? I'm assuming ALSA and Pulse are properly supported there and this can be closed with "Solved by UXP"?
